crates/runtime/src/embeddings/table.rs

```diff
@@ -46,54 +50,213 @@ pub enum Error {}
+                }
+
+                let Some(vector_length) =
+                    Self::embedding_size_from_base_table(&column, &base_schema)
+                else {
+                    tracing::warn!("Column '{}' has embeddings in base table, but the vector length could not be determined. Ignoring column.", column);
+                    continue;
+                };
+
+                embedded_columns.insert(
+                    column,
+                    EmbeddingColumnConfig {
+                        model_name: model,
+                        vector_size: vector_length,
+                        in_base_table: true,
+                        chunker: None, // Don't need chunking since it is done in base table.
+                    },
+                );
+            } else {
+                tracing::debug!("Column '{column}' does not have needed embeddings in base table. Will augment with model {model}.");
+
+                let Some(vector_length) =
+                    Self::embedding_size_from_models(&model, &embedding_models).await
+                else {
+                    tracing::warn!("For column '{column}', cannot precompute vector length from model '{model}'. Ignoring column.");
+                    continue;
+                };
+
+                let mut chunker = None;
+                if let Some(chunking_config) = chunking_config_opt {
+                    chunker =
+                        Self::construct_chunker(&model, chunking_config, &embedding_models).await;
+                    if chunker.is_none() {
+                        tracing::warn!("Column '{}' expects to be chunked, but the model '{}' does not support chunking. Ignoring chunking config.", column, model);
+                    }
+                }
+
+                embedded_columns.insert(
+                    column,
+                    EmbeddingColumnConfig {
+                        model_name: model,
+                        vector_size: vector_length,
+                        in_base_table: false,
+                        chunker,
+                    },
+                );
+            }
+        }
 
         Self {
             base_table,
             embedded_columns,
             embedding_models,
-            embedding_sizes: sizes,
-            embedding_chunkers: chunkers,
         }
     }
 
+    /// Check if the base table has a column that is augmented with an embedding.
+    /// For a base table with column, c, we expect:
+    ///  - `c` to be in the base schema.
+    ///  - `c_embedding` to be in the base schema. It needs to have a type compatible with [`Self::embedding_fields`].
+    ///  - If `c_embedding` has a doubly-nested list type, `c_offsets` should also be in the base schema. It should be a `List[FixedSizeList[Int32, 2]]`.
+    fn base_table_has_embedding_column(base_schema: &SchemaRef, column: &str) -> bool {
```

Comment:
This function could use some tests

Reply:
It does! 
https://github.com/spiceai/spiceai/pull/3045/files#diff-8d5f8be44e4ee0a1cfd45890df5968e4a98ff28c35190db9e6377f80f59845f7R567